### ISSUE TITLE
Derived Types with recursive definition

### ISSUE THREAD

**Issue by gkenway** [jameskermode/f90wrap#72]
I'm in the process of wrapping a Fortran code that uses a derived type within the same the derived type, ie. it is recursive. This causes problems when the wrapper is written as there are two definitions of the same type. 

I've written a simple example that replicates the issue. 
[f90wrap_test.tar.gz](https://github.com/jameskermode/f90wrap/files/1637215/f90wrap_test.tar.gz)

This should give the error:
```
gfortran:f90: f90wrap_tree_test.f90
f90wrap_tree_test.f90:11.22:

    type node_ptr_type
                      1
Error: Derived type definition of 'node_ptr_type' at (1) has already been defined
```
I've found that it can be fixed by modifying the line in the _write_scalar_wrapper routine in the f90wrapgen.py file. Replace:

```
if el.type.startswith('type'):
    self.write_type_lines(el.type)
```
with
```
if el.type.startswith('type') and (ft.strip_type(t.name) != ft.strip_type(el.type)):
    self.write_type_lines(el.type)
```
There may be other places where this would be an issue, but this seemed to fix the scalar case for me at least.

**Comment by jameskermode** [comment]
Thanks for reporting - your fix looks reasonably general to me, would you be willing to put together a pull request with this change made and your test case added as a new test case in a subdirectory of `examples/`?

**Comment by gkenway** [comment]
Absolutely. I'll do that this week.

**Comment by jameskermode** [comment]
Closed by #74
